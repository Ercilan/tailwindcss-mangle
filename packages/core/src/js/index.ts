import type { StringLiteral, TemplateElement } from '@babel/types'
import MagicString from 'magic-string'
import { jsStringEscape } from '@ast-core/escape'
import type { IJsHandlerOptions } from '../types'
import { makeRegex, splitCode } from '../shared'
import { parse, traverse } from '@/babel'

export { preProcessJs, preProcessRawCode } from './pre'

export function handleValue(raw: string, node: StringLiteral | TemplateElement, options: IJsHandlerOptions, ms: MagicString, offset: number, escape: boolean) {
  const { ctx, splitQuote = true } = options
  const { replaceMap, classGenerator: clsGen } = ctx

  const array = splitCode(raw, {
    splitQuote,
  })
  let rawString = raw
  let needUpdate = false
  for (const v of array) {
    if (replaceMap.has(v)) {
      let ignoreFlag = false
      if (Array.isArray(node.leadingComments)) {
        ignoreFlag = node.leadingComments.findIndex(x => x.value.includes('tw-mangle') && x.value.includes('ignore')) > -1
      }

      if (!ignoreFlag) {
        rawString = rawString.replace(makeRegex(v), clsGen.generateClassName(v).name)
        needUpdate = true
      }
    }
  }
  if (needUpdate && typeof node.start === 'number' && typeof node.end === 'number') {
    const start = node.start + offset
    const end = node.end - offset

    if (start < end && raw !== rawString) {
      ms.update(start, end, escape ? jsStringEscape(rawString) : rawString)
    }
  }
  return rawString
}

export function jsHandler(rawSource: string | MagicString, options: IJsHandlerOptions) {
  const ms: MagicString = typeof rawSource === 'string' ? new MagicString(rawSource) : rawSource
  const ast = parse(ms.original, {
    sourceType: 'unambiguous',
    plugins: ['typescript', 'jsx'],
  })
  traverse(ast, {
    StringLiteral: {
      enter(p) {
        const n = p.node
        handleValue(n.value, n, options, ms, 1, true)
      },
    },
    TemplateElement: {
      enter(p) {
        const n = p.node
        handleValue(n.value.raw, n, options, ms, 0, false)
      },
    },
    // CallExpression: {
    //   enter(p: NodePath<CallExpression>) {
    //     const calleePath = p.get('callee')
    //     if (calleePath.isIdentifier() && calleePath.node.name === 'eval') {
    //       p.traverse({
    //         StringLiteral: {
    //           enter(s) {
    //             // ___CSS_LOADER_EXPORT___
    //             const res = jsHandler(s.node.value, options)
    //             if (res.code) {
    //               s.node.value = res.code
    //             }
    //           },
    //         },
    //       })
    //     }
    //   },
    // },
  })

  return {
    code: ms.toString(),
    get map() {
      return ms.generateMap()
    },
  }
}
