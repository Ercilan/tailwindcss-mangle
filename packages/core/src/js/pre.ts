import babel from '@babel/core'
import { declare } from '@babel/helper-plugin-utils'
import MagicString from 'magic-string'
import { splitCode } from '@tailwindcss-mangle/shared'
import { sort } from 'fast-sort'
import { jsStringEscape } from '@ast-core/escape'
import { getStringLiteralCalleeName, getTemplateElementCalleeName } from './utils'
import type { Context } from '@/ctx'

interface Options {
  replaceMap: Map<string, string>
  magicString: MagicString
  id: string
  ctx: Context
}

type HandleValueOptions = {
  raw: string
  path: babel.NodePath<babel.types.StringLiteral | babel.types.TemplateElement>
  offset: number
  escape: boolean
  preserve: boolean
} & Options

export function handleValue(options: HandleValueOptions) {
  const { ctx, id, path, magicString, raw, replaceMap, offset = 0, escape = false, preserve = false } = options
  const node = path.node
  let value = raw
  const arr = sort(splitCode(value)).desc((x) => x.length)

  for (const str of arr) {
    if (replaceMap.has(str)) {
      ctx.addToUsedBy(str, id)
      if (preserve) {
        ctx.addPreserveClass(str)
      }
      // replace
      const v = replaceMap.get(str)
      if (v) {
        value = value.replaceAll(str, v)
      }
    }
  }
  if (preserve) {
    return
  }
  if (typeof node.start === 'number' && typeof node.end === 'number' && value) {
    const start = node.start + offset
    const end = node.end - offset
    if (start < end) {
      magicString.update(start, end, escape ? jsStringEscape(value) : value)
    }
  }
}

export const plugin = declare((api, options: Options) => {
  api.assertVersion(7)
  const { magicString, replaceMap, id, ctx } = options
  return {
    visitor: {
      StringLiteral: {
        enter(p) {
          const opts: HandleValueOptions = {
            ctx,
            id,
            magicString,
            path: p,
            raw: p.node.value,
            replaceMap,
            offset: 1,
            escape: true,
            preserve: false
          }
          const calleeName = getStringLiteralCalleeName(p)
          if (calleeName && ctx.isPreserveFunction(calleeName)) {
            opts.preserve = true
          }
          handleValue(opts)
        }
      },
      TemplateElement: {
        enter(p) {
          const opts: HandleValueOptions = {
            ctx,
            id,
            magicString,
            path: p,
            raw: p.node.value.raw,
            replaceMap,
            offset: 0,
            escape: false,
            preserve: false
          }
          const calleeName = getTemplateElementCalleeName(p)
          if (calleeName && ctx.isPreserveFunction(calleeName)) {
            opts.preserve = true
          }
          handleValue(opts)
        }
      }
    }
  }
})

export function preProcessJs(options: { code: string | MagicString; replaceMap: Map<string, string>; id: string; ctx: Context }) {
  const { code, replaceMap, id, ctx } = options
  const magicString = typeof code === 'string' ? new MagicString(code) : code

  babel.transformSync(magicString.original, {
    presets: [
      // ['@babel/preset-react', {}],
      [
        require('@babel/preset-typescript'),
        {
          allExtensions: true,
          isTSX: true
        }
      ]
    ],
    plugins: [
      [
        plugin,
        {
          magicString,
          replaceMap,
          id,
          ctx
        }
      ]
    ],
    filename: id
  })
  return magicString.toString()
}
