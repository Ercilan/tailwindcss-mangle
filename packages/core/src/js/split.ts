export const validateFilterRE = /[\w\u00A0-\uFFFF-_:%-?]/

export function isValidSelector(selector = ''): selector is string {
  return validateFilterRE.test(selector)
}

export const splitCode = (
  code: string,
  options: {
    splitQuote?: boolean
  } = { splitQuote: true }
) => {
  const regex = options.splitQuote ? /[\s"]+/ : /[\s]+/
  return code.split(regex).filter(isValidSelector)
}
