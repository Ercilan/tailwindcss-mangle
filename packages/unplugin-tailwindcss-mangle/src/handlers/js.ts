import { parse, traverse, generate } from '../lib/babel'

import type { Node } from '@babel/types'
import type { TraverseOptions, IJsHandlerOptions } from '../types'
import { escapeStringRegexp } from '../utils'
import { splitCode } from './split'

export function jsHandler(rawSource: string, options: IJsHandlerOptions) {
  const ast = parse(rawSource)
  const set = options.set
  const clsGen = options.classGenerator
  // 这样搞会把原先所有的 children 含有相关的 也都转义了
  const topt: TraverseOptions<Node> = {
    StringLiteral: {
      enter(p) {
        const n = p.node
        const arr = splitCode(n.value) // .split(/\s/).filter((x) => x)
        let rawStr = n.value
        for (let i = 0; i < arr.length; i++) {
          const v = arr[i]
          if (set.has(v)) {
            let ignoreFlag = false
            if (Array.isArray(n.leadingComments)) {
              ignoreFlag = n.leadingComments.findIndex((x) => x.value.includes('tw-mangle') && x.value.includes('ignore')) > -1
            }

            if (!ignoreFlag) {
              rawStr = rawStr.replace(new RegExp(escapeStringRegexp(v), 'g'), clsGen.generateClassName(v).name)
            }
          }
        }
        n.value = rawStr
      }
    },
    noScope: true
  }

  traverse(ast, topt)

  return generate(ast)
}
