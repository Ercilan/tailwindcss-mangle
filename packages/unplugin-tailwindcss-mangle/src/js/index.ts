import _generate from '@babel/generator'
import { parse } from '@babel/parser'
import _traverse from '@babel/traverse'

import type { Node, StringLiteral, TemplateElement } from '@babel/types'
import type { TraverseOptions, IHandlerOptions } from '../types'
import { escapeStringRegexp } from '../utils'
import { splitCode } from './split'

function getDefaultExportFromNamespaceIfPresent(n: any) {
  return n && Object.prototype.hasOwnProperty.call(n, 'default') ? n.default : n
}
const generate = getDefaultExportFromNamespaceIfPresent(_generate) as typeof _generate
const traverse = getDefaultExportFromNamespaceIfPresent(_traverse) as typeof _traverse

export function makeRegex(str: string) {
  return new RegExp('(?<=^|[\\s"])' + escapeStringRegexp(str), 'g')
}

export function handleValue(str: string, node: StringLiteral | TemplateElement, options: IHandlerOptions) {
  const set = options.runtimeSet
  const clsGen = options.classGenerator
  const arr = splitCode(str)
  let rawStr = str
  for (let i = 0; i < arr.length; i++) {
    const v = arr[i]
    if (set.has(v)) {
      let ignoreFlag = false
      if (Array.isArray(node.leadingComments)) {
        ignoreFlag = node.leadingComments.findIndex((x) => x.value.includes('tw-mangle') && x.value.includes('ignore')) > -1
      }

      if (!ignoreFlag) {
        rawStr = rawStr.replace(makeRegex(v), clsGen.generateClassName(v).name)
      }
    }
  }
  return rawStr
}

export function jsHandler(rawSource: string, options: IHandlerOptions) {
  const ast = parse(rawSource)

  const topt: TraverseOptions<Node> = {
    StringLiteral: {
      enter(p) {
        const n = p.node
        n.value = handleValue(n.value, n, options)
      }
    },
    TemplateElement: {
      enter(p) {
        const n = p.node
        n.value.raw = handleValue(n.value.raw, n, options)
      }
    },
    noScope: true
  }

  traverse(ast, topt)

  return generate(ast)
}
