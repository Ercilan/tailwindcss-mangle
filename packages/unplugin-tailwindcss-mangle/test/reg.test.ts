// import { escapeStringRegexp } from '../src/utils'
import { makeRegex } from '../src/js/index'
describe('regex', () => {
  it('z-10 regex', () => {
    const testCase = 'z-10 w-full max-w-5xl items-center justify-between font-mono text-sm lg:flex'
    const regex = makeRegex('z-10')
    const arr = Array.from(testCase.matchAll(regex))
    expect(arr.length).toBe(1)
    expect(arr).toMatchSnapshot()
  })
})
