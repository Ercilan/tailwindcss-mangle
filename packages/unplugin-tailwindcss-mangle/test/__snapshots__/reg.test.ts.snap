// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`regex z-10 regex 1`] = `
[
  [
    "z-10",
  ],
]
`;
